File: app/schemas/qa.py
from __future__ import annotations

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.source_locator import (
    PreviewTargetRead,
    SourceLocatorRead,
    normalize_locator_fields,
)


class QuestionAnswerRequest(BaseModel):
    question: str = Field(min_length=1, max_length=4000)
    top_k: int = Field(default=5, ge=1, le=20)
    conversation_id: int | None = Field(default=None, ge=1)
    mode: str = "auto"

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Question cannot be empty.")
        return normalized

    @field_validator("mode")
    @classmethod
    def normalize_mode(cls, value: str) -> str:
        normalized = value.strip().lower()
        supported = {"auto", "chunk_only", "overview", "graph_vector_mix", "hybrid_text"}
        if normalized not in supported:
            raise ValueError("Unsupported retrieval mode.")
        return normalized


class CitationRead(BaseModel):
    citation_id: int | None = None
    citation_marker: str | None = None
    citation_unit_id: int | None = None
    chunk_db_id: int | None = None
    chunk_id: str
    document_id: int
    knowledge_base_id: int
    scope: str
    team_id: int | None
    document_name: str | None = None
    snippet: str | None = None
    text: str
    source_type: str | None = None
    char_start: int | None = None
    char_end: int | None = None
    page_number: int | None = None
    start_time: float | None = None
    end_time: float | None = None
    section_title: str | None = None
    source_locator: SourceLocatorRead | None = None
    preview_target: PreviewTargetRead | None = None
    heading_path: list[str] = Field(default_factory=list)
    citation_ready: bool = False
    retrieval_mode: str | None = None
    score: float | None = None

    @model_validator(mode="before")
    @classmethod
    def normalize_legacy_source_locator(cls, value: object) -> object:
        normalized = normalize_locator_fields(value)
        if not isinstance(normalized, dict):
            return normalized

        payload = dict(normalized)
        payload["heading_path"] = _normalize_heading_path(payload.get("heading_path"))
        payload["char_start"], payload["char_end"] = _normalize_char_range(
            payload.get("char_start"),
            payload.get("char_end"),
        )
        payload["source_locator"] = _normalize_nested_locator(
            payload.get("source_locator"),
            normalize_heading=True,
        )
        payload["preview_target"] = _normalize_nested_locator(
            payload.get("preview_target"),
            normalize_heading=False,
        )
        if "citation_ready" not in payload:
            payload["citation_ready"] = (
                payload.get("citation_unit_id") is not None
                and payload.get("source_locator") is not None
            )
        return payload


def _normalize_heading_path(value: object) -> list[str]:
    if isinstance(value, str):
        normalized = value.strip()
        return [normalized] if normalized else []
    if not isinstance(value, (list, tuple)):
        return []
    return [item.strip() for item in value if isinstance(item, str) and item.strip()]


def _normalize_char_range(start: object, end: object) -> tuple[int | None, int | None]:
    if (
        isinstance(start, int)
        and not isinstance(start, bool)
        and isinstance(end, int)
        and not isinstance(end, bool)
        and 0 <= start < end
    ):
        return start, end
    return None, None


def _normalize_nested_locator(
    value: object,
    *,
    normalize_heading: bool,
) -> object:
    if isinstance(value, BaseModel):
        value = value.model_dump()
    if not isinstance(value, dict):
        return value
    normalized = dict(value)
    normalized["char_start"], normalized["char_end"] = _normalize_char_range(
        normalized.get("char_start"),
        normalized.get("char_end"),
    )
    if normalize_heading:
        normalized["heading_path"] = _normalize_heading_path(
            normalized.get("heading_path")
        )
    return normalized


class QuestionAnswerResponse(BaseModel):
    conversation_id: int
    answer: str
    citations: list[CitationRead]
    intent: str | None = None
    retrieval_mode: str | None = None
    requested_mode: str | None = None
    selected_mode: str | None = None
    router_reason: str | None = None
    used_reranker: bool | None = None
    trace_id: int | str | None = None

File: frontend/schemas/qa.ts
import { z } from "zod";

export const retrievalModeSchema = z.enum([
  "auto",
  "chunk_only",
  "overview",
  "graph_vector_mix",
  "hybrid_text"
]);

const nullableString = z.string().nullish().default(null);
const nullableNumber = z.number().finite().nullish().default(null);
const nullableId = z.number().int().positive().nullish().default(null);
const headingPathSchema = z.preprocess(
  (value) => value ?? [],
  z.array(z.string())
);

const sourceLocatorSchema = z.object({
  kind: z.enum(["text_range", "pdf_page", "image_region", "time_range", "unknown"]),
  document_id: nullableId,
  source_type: nullableString,
  source_locator_text: nullableString,
  char_start: nullableNumber,
  char_end: nullableNumber,
  section_title: nullableString,
  heading_path: headingPathSchema,
  page_number: nullableNumber,
  page_region: z.record(z.unknown()).nullish().default(null),
  bbox: z.record(z.unknown()).nullish().default(null),
  region_hint: nullableString,
  ocr_provider: nullableString,
  start_time: nullableNumber,
  end_time: nullableNumber
}).superRefine(validateCharRangePair);

const previewTargetSchema = z.object({
  kind: z.literal("document_preview"),
  document_id: nullableId,
  source_type: nullableString,
  locator_kind: z.enum(["text_range", "pdf_page", "image_region", "time_range", "unknown"]),
  source_locator_text: nullableString,
  char_start: nullableNumber,
  char_end: nullableNumber,
  section_title: nullableString,
  page_number: nullableNumber,
  start_time: nullableNumber,
  end_time: nullableNumber
}).superRefine(validateCharRangePair);

export const citationSchema = z.object({
  citation_id: nullableId,
  citation_marker: nullableString,
  citation_unit_id: nullableId,
  chunk_db_id: nullableId,
  chunk_id: z.string().nullish().default(null),
  document_id: nullableId,
  knowledge_base_id: nullableId,
  scope: nullableString,
  team_id: nullableId,
  document_name: nullableString,
  snippet: nullableString,
  text: z.string(),
  source_type: nullableString,
  char_start: nullableNumber,
  char_end: nullableNumber,
  page_number: nullableNumber,
  start_time: nullableNumber,
  end_time: nullableNumber,
  section_title: nullableString,
  source_locator: sourceLocatorSchema.nullish().default(null),
  preview_target: previewTargetSchema.nullish().default(null),
  heading_path: headingPathSchema,
  citation_ready: z.boolean().optional().default(false),
  retrieval_mode: retrievalModeSchema.nullish().default(null),
  score: nullableNumber
}).superRefine(validateCharRangePair);

export const askResponseSchema = z.object({
  conversation_id: z.number().int().positive(),
  answer: z.string(),
  citations: z.array(citationSchema),
  intent: nullableString,
  retrieval_mode: retrievalModeSchema.nullish().default(null),
  requested_mode: retrievalModeSchema.nullish().default(null),
  selected_mode: retrievalModeSchema.nullish().default(null),
  router_reason: nullableString,
  used_reranker: z.boolean().nullish().default(null),
  trace_id: z.union([z.number(), z.string()]).nullish().default(null)
});

export const retrievalSchema = z.object({
  query: z.string().min(1, "Query is required."),
  top_k: z.coerce.number().min(1).max(20).default(5),
  mode: retrievalModeSchema.default("auto")
});

export const askSchema = z.object({
  question: z.string().min(1, "Question is required."),
  top_k: z.coerce.number().min(1).max(20).default(5),
  conversation_id: z.number().nullable().optional(),
  mode: retrievalModeSchema.default("auto")
});

function validateCharRangePair(
  value: { char_start?: number | null; char_end?: number | null },
  context: z.RefinementCtx
) {
  const hasStart = typeof value.char_start === "number";
  const hasEnd = typeof value.char_end === "number";
  if (hasStart !== hasEnd || (hasStart && hasEnd && !(value.char_start! >= 0 && value.char_start! < value.char_end!))) {
    context.addIssue({
      code: z.ZodIssueCode.custom,
      message: "char_start and char_end must form a valid range."
    });
  }
}

export type RetrievalValues = z.infer<typeof retrievalSchema>;
export type AskValues = z.infer<typeof askSchema>;
export type CitationContract = z.infer<typeof citationSchema>;
export type AskResponseContract = z.infer<typeof askResponseSchema>;
